utils/oc.py

```diff
@@ -213,18 +214,64 @@ def recycle_pods(self, dry_run, namespace, dep_kind, dep_resource):
         pods = self.get(namespace, 'Pods')['items']
 
         if dep_kind == 'Secret':
-            pods_to_recycle = [pod['metadata']['name'] for pod in pods
+            pods_to_recycle = [pod for pod in pods
                                if self.secret_used_in_pod(dep_name, pod)]
         else:
             raise RecyclePodsUnsupportedKindError(dep_kind)
 
+        recyclables = {}
         for pod in pods_to_recycle:
-            logging.info(['recycle_pod', namespace, pod])
-            if not dry_run:
-                self.delete(namespace, 'Pod', pod)
-                logging.info(['validating_pods', namespace])
-                self.validate_pods_ready(
-                    namespace, self.secret_used_in_pod, dep_name)
+            if pod['status']['phase'] == 'Succeeded':
+                # don't try to recycle a pod that completed successfully
+                continue
+            owner = self.get_obj_root_owner(namespace, pod)
+            kind = owner['kind']
+            if kind == 'ReplicationController':
+                # need to act on the pod for replication controllers
+                owner = pod
+            recyclables.setdefault(kind, [])
+            exists = False
+            for obj in recyclables[kind]:
+                owner_name = owner['metadata']['name']
+                if obj['metadata']['name'] == owner_name:
+                    exists = True
+                    break
+            if not exists:
+                recyclables[kind].append(owner)
+
+        supported_recyclables = ['ReplicationController', 'Deployment',
+                                 'DeploymentConfig']
+        for kind, objs in recyclables.items():
+            for obj in objs:
+                if kind not in supported_recyclables:
+                    continue
+                name = obj['metadata']['name']
+                logging.info([f'recycle_{kind.lower()}', namespace, name])
+                if not dry_run:
+                    now = datetime.now()
+                    recycle_time = now.strftime("%d/%m/%Y %H:%M:%S")
+                    if kind == 'ReplicationController':
+                        self.delete(namespace, 'Pod', name)
+                        logging.info(['validating_pods', namespace])
+                        self.validate_pods_ready(
+                            namespace, self.secret_used_in_pod, dep_name,
+                            recycle_time)
```

Comment:
I think that gets us into a difficult space.  The deployments/dc have the logic, or should, to ensure a proper upgrade via their strategy.  They can appropriately wait and look at readiness probes to know if the replaced pod is functioning properly before retiring an old one.  That can take quite some time.  I don't think we want to try to duplicate that logic nor try to wait for a number of running pods because I feel our wait logic will be flawed due to timeouts, pod miscounts (we want 30 pods but only recycle 3 at a time, how do we know all 30 have recycled?).  IMO, better leave the deployment/dc handle it and ensure there's no downtime.

Reply:
ok, makes sense.